be/src/vec/runtime/shared_hash_table_controller.cpp

```diff
@@ -0,0 +1,95 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#include "shared_hash_table_controller.h"
+
+#include <runtime/runtime_state.h>
+
+namespace doris {
+namespace vectorized {
+
+bool SharedHashTableController::should_build_hash_table(RuntimeState* state, int my_node_id) {
+    std::lock_guard<std::mutex> lock(_mutex);
+    auto it = _builder_fragment_ids.find(my_node_id);
+    if (it == _builder_fragment_ids.cend()) {
+        _builder_fragment_ids[my_node_id] = state->fragment_instance_id();
+        return true;
+    }
+    return false;
+}
+
+void SharedHashTableController::put_hash_table(SharedHashTableEntry&& entry, int my_node_id) {
+    std::lock_guard<std::mutex> lock(_mutex);
+    DCHECK(_hash_table_entries.find(my_node_id) == _hash_table_entries.cend());
+    _hash_table_entries.insert({my_node_id, std::move(entry)});
+    _cv.notify_all();
+}
+
+SharedHashTableEntry& SharedHashTableController::wait_for_hash_table(int my_node_id) {
+    std::unique_lock<std::mutex> lock(_mutex);
+    auto it = _hash_table_entries.find(my_node_id);
+    if (it == _hash_table_entries.cend()) {
+        _cv.wait(lock, [this, &it, my_node_id]() {
```

Comment:
maybe `dead lock` when the node should be build hash table is canceled. it's dangerous